src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1027,69 +1025,321 @@ private static void ParseStatusLine(ReadOnlySpan<byte> line, HttpResponseMessage
+                    if ((colIdx == 1 && buffer[0] == '\r') || colIdx == 0)
+                    {
+                        return (finished: true, bytesConsumed: originalBufferLength - buffer.Length + colIdx + 1);
+                    }
+
+                    buffer = buffer.Slice(0, colIdx);
+                    throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_line, Encoding.ASCII.GetString(buffer)));
+                }
+
+                int lfIdx = colIdx + 1;
+                if (lfIdx >= buffer.Length) goto needMore;
+
+                Span<byte> valueStart = buffer.Slice(lfIdx);
+                Span<byte> valueIter = valueStart;
+
+                while (true)
+                {
+                    lfIdx = valueIter.IndexOf((byte)'\n');
+                    if (lfIdx < 0) goto needMore;
+
+                    int crOrLfIdx = lfIdx - 1;
+                    if (crOrLfIdx < 0 || valueIter[crOrLfIdx] != '\r')
+                    {
+                        crOrLfIdx = lfIdx;
+                    }
+
+                    int spIdx = lfIdx + 1;
+                    if (spIdx >= valueIter.Length) goto needMore;
+
+                    if (valueIter[spIdx] is not (byte)'\t' and not (byte)' ')
+                    {
+                        // Found the end of the header value.
+
+                        if (response is not null)
+                        {
+                            ReadOnlySpan<byte> headerName = buffer.Slice(0, colIdx);
+                            ReadOnlySpan<byte> headerValue = valueStart.Slice(0, valueStart.Length - valueIter.Length + crOrLfIdx);
+                            AddResponseHeader(headerName, headerValue, response, isFromTrailer);
+                        }
+
+                        buffer = valueStart.Slice(valueStart.Length - valueIter.Length + spIdx);
+                        break;
+                    }
+
+                    // Found an obs-fold (CRLFHT/CRLFSP).
+                    // Replace with SPSPSP and keep looking for the final newline.
+                    // Also handles LFHT and LFSP.
+
+                    valueIter[crOrLfIdx] = (byte)' ';
+                    valueIter[lfIdx] = (byte)' ';
+                    valueIter[spIdx] = (byte)' ';
+
+                    valueIter = valueIter.Slice(spIdx + 1);
                 }
             }
 
-            if (pos == 0)
+        needMore:
+            return (finished: false, bytesConsumed: originalBufferLength - buffer.Length);
+        }
+
+        /// <remarks>
+        /// This method REQUIRES (32-1) bytes of valid address space after <paramref name="buffer"/>.
+        /// This is done via <see cref="SimdPaddedArray"/>.
+        /// It also assumes that it can always step backwards to a 32-byte aligned address.
+        /// </remarks>
+        private unsafe (bool finished, int bytesConsumed) ParseHeadersAvx2(Span<byte> buffer, HttpResponseMessage? response, bool isFromTrailer)
```

Comment:
I think this should be changed to be `Vector128<T>` based, possibly even waiting for https://github.com/dotnet/runtime/pull/61649 to be merged and using the new xplat HWIntrinsic APIs.

We shouldn't be merging acceleration here, or anywhere else in the BCL without also including ARM64 support and that means using `Vector128<T>` as a first class path.

There may be additional optimization opportunities involving `Vector256<T>` (and therefore AVX2 or eventually SVE on ARM); but we always need a `V128<T>` path on these APIs.

CC. @danmoseley

Reply:
Thanks for this work @scalablecory ! I agree with @tannergooding that we should nowadays consider Arm64 support as essential. But doing it explicitly would be wasted work. @tannergooding how soon can `Vector128<T>` be useable ?

Reply:
(It'd also be really helpful if, as part of making it fully usable, we aggressively fixed up all current usage of intrinsics in the repo to use Vector128.  I believe that's part of the plan, just making sure.  Then subsequent usage can use that as a guide.)

Reply:
> we aggressively fixed up all current usage of intrinsics in the repo to use Vector128. I believe that's part of the plan, just making sure. 

Right, that is the current plan. `Vector128<T>` is the "common" surface area to all currently supported platforms (x86, x64, ARM64) and will likewise be supported on future platforms if we add them (WASM, ARM32, PowerPC, etc).
-- A large reason for this is because `Vector128<float>` is the natural type for graphical and multimedia based applications, so almost all CPUs add some sort of SIMD support for 128-bit vectors.

There may still be cases where having an additional path supporting `Vector256<T>` is beneficial, particularly if the workloads are known to be big. But given that `ARM32/64` doesn't support `Vector256<T>` (we could only emulate it as `2x Vector128<T>` ops), `Vector128<T>` is likely the base type that all SIMD algorithms need to start with.
-- This also ensures platforms without AVX/AVX2 support are accelerated. x64 emulation on ARM64 for example only supports up to `SSE4.1` on both Windows and MacOS. It also ensures low power or budget CPUs (such as `Intel Atom`) are supported.

> But doing it explicitly would be wasted work. @tannergooding how soon can Vector128<T> be useable ?

https://github.com/dotnet/runtime/pull/61649 is the last "big" stepping stone and ensures `Vector128<T>.IsHardwareAccelerated` returns `true`. It's just pending review from @echesakovMSFT (who I pinged this morning and indicated they would try to get to it this week).
-- There are still a couple additional PRs to go up after this, but they mainly are for adding `nint/nuint` support and adding in a couple of additional APIs that were approved for after the initial API review for the xplat APIs, such as the `Sum` method.

Reply:
> We shouldn't be merging acceleration here, or anywhere else in the BCL without also including ARM64 support and that means using `Vector128<T>` as a first class path.

Can you expand here? I can read this as three ways:

- We have a new cross-plat vector type that we think will be applicable here, but it isn't complete, and maybe we can use this as a test case.
- We have red tape saying no platform-specific intrinsics without also having an ARM64 path. This seems like we'd be letting perfect get in the way of good -- I'm happy to revisit an ARM64 version in a separate issue, but delaying this change doesn't seem to add any value?
- We intend to do away with platform-specific intrinsics in libraries so we shouldn't start adding more.

Reply:
> We have red tape saying no platform-specific intrinsics without also having an ARM64 path.

I think this is the most applicable. ARM64, particularly for .NET 7, is a major top down focus. It is a first class platform and we should not be adding new code that accelerates x64 and skips out on accelerating ARM64.

> We have new cross-plat vector stuff that we think will be applicable here, but it isn't complete, 

The new cross-platform vector APIs will be applicable here. It's practically ready, just pending the last PR getting merged. The entire purpose of it is to help trivialize support ARM64 (and other non-x64 platforms) because we have a goal of ensure they are also treated as "first class".

In most of the cases where we have SIMD code supporting both `x64` and `ARM64`, the two code paths are nearly identical. Often only differing in a couple places, if anywhere at all. The new APIs allow you to do things like `x + y` rather than `AdvSimd.Add(x, y)` and `Sse.Add(x, y)`. This allows most of the code to be shared.

There will still be a handful of places where there is no common functionality available or where there it is advantageous to specialize on a given platform. For those scenarios, you then trivially switch from `xplat` intrinsics to `hardware specific intrinsics` using `if (AdvSimd.IsSupported) { /* ARM64 path */ } else if (Sse.IsSupported) { /* x86/x64 path */ else { /* fallback path */ }`. Part of the reason this is "trivial" is because you are already using `Vector128<T>` and so you have `shared simd logic; plat-specific simd logic; shared simd logic`.

> We intend to do away with platform-specific intrinsics in libraries so we shouldn't start adding more.

It won't go away, and there will still be places where there are platform-specific optimization opportunities available. But by default, most code adding SIMD support will have a `Vector128<T>` shared path that works for `ARM64` and for small inputs on `x86/x64`. This same path will also allow implicit lightup on future platforms getting SIMD support, such as `WASM`.

Reply:
Thank you for the clarification.

> There will still be a handful of places where there is no common functionality available or where there it is advantageous to specialize on a given platform. For those scenarios, you then trivially switch from `xplat` intrinsics to `hardware specific intrinsics` using `if (AdvSimd.IsSupported) { /* ARM64 path */ } else if (Sse.IsSupported) { /* x86/x64 path */ else { /* fallback path */ }`. Part of the reason this is "trivial" is because you are already using `Vector128<T>` and so you have `shared simd logic; plat-specific simd logic; shared simd logic`.

This seems very reasonable.

> > We have red tape saying no platform-specific intrinsics without also having an ARM64 path.
> 
> I think this is the most applicable. ARM64, particularly for .NET 7, is a major top down focus. It is a first class platform and we should not be adding new code that accelerates x64 and skips out on accelerating ARM64.

Thanks, this makes sense. I agree fully with the goal but we need to work on how we word this guidance (for docs, and future PR feedback).

We shouldn't state it as a dogma of no Vector256, but rather something like "here's how we want cross-plat vector code to work (see your great fallback example); PRs should prefer this were possible, and justify it if not."